### PR TITLE
bpo-31412: wave.open takes a path-like object

### PR DESCRIPTION
**Pull request by mscuthbert**
https://bugs.python.org/issue31412

### CODE REVIEW

File: Lib/wave.py [hunk 1]
```diff
@@ -495,6 +499,13 @@ def open(f, mode=None):
             mode = f.mode
         else:
             mode = 'rb'
+    try:
+        f = os.fspath(f)
+    except TypeError:
+        if not hasattr(f, 'read') or not hasattr(f, 'close'):
```
Comment (berkerpeksag) [inline Lib/wave.py]: I'd also check for a `write` method here.

File: Lib/wave.py [hunk 2]
```diff
@@ -495,6 +499,13 @@ def open(f, mode=None):
...
+            raise TypeError('open() takes str, bytes, a PathLike object, '
+                            + f'or an open filehandle, not {type(f)}')
```
Comment (berkerpeksag) [inline Lib/wave.py]: `{type(f).__name__!r}` would give us a more readable name.

File: Misc/NEWS.d/next/Library/2017-09-10-15-20-14.bpo-31412.lNbm-A.rst [hunk 3]
```diff
@@ -0,0 +1,2 @@
+`wave.open()` now accepts PathLike objects. Contributed by Michael Scott
```
Comment (berkerpeksag) [inline Misc/NEWS.d/next/Library/2017-09-10-15-20-14.bpo-31412.lNbm-A.rst]: \`wave.open()\` -> :func:\`wave.open\`

Comment (berkerpeksag) [inline Misc/NEWS.d/next/Library/2017-09-10-15-20-14.bpo-31412.lNbm-A.rst]: PathLike -> :class:\`os.PathLike\`

Comment (berkerpeksag) [inline Misc/NEWS.d/next/Library/2017-09-10-15-20-14.bpo-31412.lNbm-A.rst]: "Contributed by" -> "Patch by"

We use the former in `Doc/whatsnew/*`.

File: Lib/wave.py [hunk 4]
```diff
@@ -1,13 +1,14 @@
-"""Stuff to parse WAVE files.
+"""Module for parsing WAVE audio files.
 
 Usage.
 
 Reading WAVE files:
       f = wave.open(file, 'r')
-where file is either the name of a file or an open file pointer.
```
Comment (berkerpeksag) [inline Lib/wave.py]: Could you create a separate PR for these changes? We've removed duplicate docstrings from several modules (e.g. the venv module) to avoid future maintenance costs. I think we can also drop this giant docstring too.

File: Lib/wave.py [hunk 5]
```diff
@@ -1,13 +1,14 @@
-"""Stuff to parse WAVE files.
+"""Module for parsing WAVE audio files.
```
Comment (berkerpeksag) [inline Lib/wave.py]: This change is out of scope for this issue.

File: Lib/test/test_wave.py [hunk 6]
```diff
@@ -2,10 +2,12 @@
 from test import audiotests
 from test import support
 from audioop import byteswap
+import io
+import pathlib
+import os
 import sys
 import wave
 
-
```
Comment (berkerpeksag) [inline Lib/test/test_wave.py]: Please revert this change.

File: Lib/test/test_wave.py [hunk 7]
```diff
@@ -108,6 +110,26 @@ def test__all__(self):
         blacklist = {'WAVE_FORMAT_PCM'}
         support.check__all__(self, wave, blacklist=blacklist)
 
+    def test_open(self):
```
Comment (berkerpeksag) [inline Lib/test/test_wave.py]: I'd rename this to something like `test_open_pathlike`.

File: Lib/wave.py (hunk 2 above)
Comment (berkerpeksag) [inline Lib/wave.py]: I think we also need to add `from None` to make the traceback clearer.

Comment (berkerpeksag) [inline Lib/wave.py]: "open filehandle" -> file-like object

**Review by berkerpeksag — CHANGES_REQUESTED** [review]

File: Lib/wave.py (hunk 1 above)
Comment (serhiy-storchaka) [reply Lib/wave.py]: I think it is enough to check for a `read` in `Wave_read` and for a `write` on `Wave_write`. `close` should be optional.

Note that in `Wave_read` and `Wave_write` constructors there are checks `isinstance(f, str)`. This means that bytes patch are not supported, and file descriptors are not supported. I would use `hasattr(f, 'read')` or `hasattr(f, 'write')` instead and just pass other objects to `builtins.open()`.

File: Lib/wave.py (hunk 2 above)
Comment (serhiy-storchaka) [reply Lib/wave.py]: Actually bytes are not supported with current implementation. See the comment above.

File: Misc/NEWS.d/next/Library/2017-09-10-15-20-14.bpo-31412.lNbm-A.rst (hunk 3 above)
Comment (serhiy-storchaka) [reply Misc/NEWS.d/next/Library/2017-09-10-15-20-14.bpo-31412.lNbm-A.rst]: Or just `path-like`. Or `:term:'path-like objects <path-like object>'` (use backquotes instead of apostrophes).

File: Doc/library/wave.rst [hunk 8]
```diff
@@ -47,6 +47,9 @@ The :mod:`wave` module defines the following function and exception:
    .. versionchanged:: 3.4
       Added support for unseekable files.
 
+   .. versionchanged:: 3.7
+      Added support for :term:`path-like object`.
```
Comment (serhiy-storchaka) [inline Doc/library/wave.rst]: Or `:term:'path-like objects <path-like object>'` (use backquotes instead of apostrophes).

File: Doc/library/wave.rst [hunk 9]
```diff
@@ -19,8 +19,8 @@ The :mod:`wave` module defines the following function and exception:
 
 .. function:: open(file, mode=None)
 
-   If *file* is a string, open the file by that name, otherwise treat it as a
-   file-like object.  *mode* can be:
+   If *file* is a string or :term:`path-like object`, open the file
+   by that name, otherwise treat it as a file-like object.  *mode* can be:
```
Comment (serhiy-storchaka) [inline Doc/library/wave.rst]: Maybe add a reference to the "file-like object" term?

File: Lib/test/test_wave.py (hunk 7 above)
Comment (serhiy-storchaka) [reply Lib/test/test_wave.py]: This test checks only failed cases. It would be better to test successful opens.

Test not only reading, but writing too.

**Review by serhiy-storchaka — CHANGES_REQUESTED** [review]
If change the `wave` module, the `aifc` and `sunau` modules should be changed too. All three module should have the same interface.

File: Lib/test/test_wave.py (hunk 7 above)
Comment (mscuthbert) [reply Lib/test/test_wave.py]: Is it acceptable to add a wave file to the test directory?  I didn't test on a .wav file since even a small one can be 20-30k -- if that's okay then I will add one and test that.

**Comment by mscuthbert** [comment]
@serhiy-storchaka  -- thanks for the comments -- would patching aifc and sunau be in the scope for this PR or open a separate issue for them?

File: Lib/test/test_wave.py (hunk 7 above)
Comment (serhiy-storchaka) [reply Lib/test/test_wave.py]: There are sample wave files in the audiodata directory. See how other tests use them.

**Comment by serhiy-storchaka** [comment]
I think it is better to do this in one issue. These modules have much common and use common tests.

**Comment by mscuthbert** [comment]
I have made the requested changes; please review again

File: Doc/library/wave.rst (hunk 9 above)
Comment (mscuthbert) [reply Doc/library/wave.rst]: I looked in the glossary and "file-like object" redirects to "file object" so I used that term instead.  Hope that that's okay.

File: Doc/library/sunau.rst [hunk 10]
```diff
@@ -58,6 +59,8 @@ The :mod:`sunau` module defines the following functions:
    A *mode* of ``'r'`` returns an :class:`AU_read` object, while a *mode* of ``'w'``
    or ``'wb'`` returns an :class:`AU_write` object.
 
+   .. versionchanged:: 3.7
```
Comment (methane) [inline Doc/library/sunau.rst]:
```suggestion
   .. versionchanged:: 3.11
```

File: Doc/library/aifc.rst [hunk 11]
```diff
@@ -50,6 +51,9 @@ Module :mod:`aifc` defines the following function:
    .. versionchanged:: 3.4
       Support for the :keyword:`with` statement was added.
 
+   .. versionchanged:: 3.7
```
Comment (methane) [inline Doc/library/aifc.rst]:
```suggestion
   .. versionchanged:: 3.11
```

**Review by methane — CHANGES_REQUESTED** [review]

**Comment by csabella** [comment]
@mscuthbert, please resolve the merge conflicts and address the code review comments.

**Comment by mscuthbert** [comment]
> @mscuthbert, please resolve the merge conflicts and address the code review comments.

Hi @csabella -- I will do the merge conflicts soon, but I don't believe there were any code review comments -- my last correspondence was waiting for a review of the other changes.  Thanks for taking the time to look at this again.

File: Doc/library/wave.rst [hunk 12]
```diff
@@ -47,6 +47,9 @@ The :mod:`wave` module defines the following function and exception:
    .. versionchanged:: 3.4
       Added support for unseekable files.
 
+   .. versionchanged:: 3.8
```
Comment (methane) [inline Doc/library/wave.rst]:
```suggestion
   .. versionchanged:: 3.11
```

**Comment by jacobtylerwalls** [comment]
> I don't believe there were any code review comments

Hi @mscuthbert , I was filtering on audio issues and stumbled across this. I've gathered "awaiting changes" means "waiting for author", and it was added when the versionchanged annotations were asked to be bumped; I bet you could bump all the way to 3.11 and then say the magic words to get the label removed and then you'd be more likely to get a substantive review based on having the right labels.

**Comment by hugovk** [comment]
Note the aifc and sunau modules are deprecated in 3.11 and set for removal in 3.13, so maybe those changes could be dropped? The wave module is being kept.

See [PEP 594 – Removing dead batteries from the standard library](https://peps.python.org/pep-0594/) and https://github.com/python/cpython/issues/91217.

**Comment by methane** [comment]
Those modules shares most tests. So I prefer to change them all, althoguh aifc and sunau are deprecated.

**Comment by AA-Turner** [comment]
@mscuthbert please can you merge master? 

A

**Comment by AA-Turner** [comment]
OK, I will close this branch and mark the issue as pending -- if you manage to get things working please ping me and I can review/update things as appropriate.

A

**Comment by arhadthedev** [comment]
Superseded by more general gh-102425.
